src/routes/api/client-credentials/microsoft/token.ts

```diff
@@ -0,0 +1,163 @@
+      : { status: e.status, body: await e.json() };
+  } else {
+    console.log(e);
+    return { status: 500 };
+  }
+};
+
+/**
+ * @param {RequestEvent} requestEvent
+ *
+ * @returns {Promise<
+ *   | { body: { query: ValidationError[] }; status: 400 }
+ *   | { body: ResponseDTO; status: 200 }
+ *   | { body: any; status: number }
+ *   | { status: 500 }
+ * >}
+ */
+export const POST = async (requestEvent: RequestEvent) => {
+  class RequestQueryDTO {
+    @Expose()
+    @IsNotEmpty()
+    @IsUUID(4)
+    public readonly activationId!: string;
+  }
+
+  class RequestHeaderDTO {
+    @Expose()
+    @IsNotEmpty()
+    @Transform(({ obj }) => obj?.authorization?.toString()?.replace(/^Bearer /i, ''))
+    public readonly authorization!: string;
+  }
+
+  class ResponseDTO implements ToJSON {
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.expires_in, { toClassOnly: true })
+    public readonly expiresIn!: number;
+
+    @Expose()
+    @Transform(({ value }) => new Date(value).toUTCString(), { toPlainOnly: true })
+    public readonly expiresAt?: string;
+
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.token_type, { toClassOnly: true })
+    public readonly tokenType!: string;
+
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.access_token, { toClassOnly: true })
+    public readonly accessToken!: string;
+
+    toJSON(): Exclude<JSONValue, ToJSON> {
+      return instanceToPlain(this);
+    }
+  }
+
+  // validate request query
+  const searchParams = Object.fromEntries(requestEvent.url.searchParams);
+  const query = plainToInstance(RequestQueryDTO, searchParams, classTransformOptions);
+  const queryValidationErrors = validateSync(query, classValidationOptions);
+  if (queryValidationErrors.length > 0) {
+    return { status: 400, body: { query: queryValidationErrors } };
+  }
+
+  // validate request header
+  const obj = humps.camelizeKeys(Object.fromEntries(requestEvent.request.headers.entries()));
+  const header = plainToInstance(RequestHeaderDTO, obj, classTransformOptions);
+  const headerValidationErrors = validateSync(header, classValidationOptions);
+  if (headerValidationErrors.length > 0) {
+    return { status: 401 };
+  }
+
+  const orm = await MikroORM.init({ ...config, ...{ entities: [Activation] } });
+  const em = orm.em.fork();
+
+  const subject = webexHttpPeopleResource(header.authorization)
+    .getMyOwnDetails()
+    .then((r) => r.json())
+    .then((r) => r.emails[0]);
+
+  const account = em.findOneOrFail(Activation, query.activationId, {
+    fields: ['microsoftTenantId', 'microsoftClientId', 'microsoftPrivateKey', 'microsoftClientCertificate']
+  });
+
+  const formJwt = (audience: string, subject: string, issuer: string, pkcs8: string, x509: string, expiry = '1h') => {
+    const alg = 'RS256';
+    const data = x509
+      .replaceAll('\n', '')
+      ?.match(/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/gm)
```

Comment:
global constant perhaps?

Reply:
Will make this a constant as part of PR https://github.com/wxsd-sales/roomos-device-widgets/pull/36.